### PR TITLE
[WIP] add explicit constructs with single arg

### PR DESCRIPTION
**Pull request by FrazerClews**
used cppcheck 1.86 to catch and add explicit to constructors with single arguments

### CODE REVIEW

File: modules/mono/utils/thread_local.h [hunk 1]
```diff
@@ -118,11 +118,11 @@ class ThreadLocal {
 		_initialize(T());
 	}
 
-	ThreadLocal(const T &p_init_val) {
+	explicit ThreadLocal(const T &p_init_val) {
 		_initialize(p_init_val);
 	}
 
-	ThreadLocal(const ThreadLocal &other) {
+	ThreadLocal(const ThreadLocal &other) { // TODO explicit?
```
Comment (neikeq) [inline modules/mono/utils/thread_local.h]: Neither of these two constructors should be explicit.

**Comment by vnen** [comment]
This is interesting but it's pointless without fixing the compiling errors it causes.

Also, I don't think `String(const char*)` should be explicit, it just makes a hassle writing code. Maybe some others as well, I didn't look at everything.

File: modules/mono/utils/thread_local.h (hunk 1 above)
Comment (FrazerClews) [reply modules/mono/utils/thread_local.h]: just double checking as cppcheck didnt pick them up, definitely not a perfect program, but i am using an older version annoyingly, will fix

**Comment by FrazerClews** [comment]
@vnen could be another issue for another PR to solve at somepoint, but can be removed for now?

File: modules/mono/utils/thread_local.h (hunk 1 above)
Comment (neikeq) [reply modules/mono/utils/thread_local.h]: We have the macro `_THREAD_LOCAL_` that can be used as:

```C++
_THREAD_LOCAL_(int) foo = 20;

// If thread_local is supported, expands to...
thread_local int foo = 20;

// If neither thread_local, __thread or __declspec((thread) are supported
ThreadLocal<int> foo = 20; // needs implicit ThreadLocal(const T &p_init_val)
```

**Comment by vnen** [comment]
> @vnen could be another issue for another PR to solve at somepoint, but can be removed for now?

I don't see the point of removing just to add it later, it's double work. We need to define what can be explicit and what not and merge only the needed changes.

The Variant ones also can implicit IMO.

In any case, I think you can stop working on this until we have feedback from the main devs when they come back from the event they're attending.

**Comment by neikeq** [comment]
> The Variant ones also can implicit IMO.

Yes, they should not be explicit. One should be able to do `Variant v = String();`, for example.

This should be reviewed carefully to make sure there are no other similar cases.

**Comment by FrazerClews** [comment]
ok, ill do the improvements you suggested and push that up and leave it from there

**Comment by aaronfranke** [comment]
Does this PR actually solve any specific problems, or is it just trying to improve internal code quality?

**Comment by FrazerClews** [comment]
> Does this PR actually solve any specific problems, or is it just trying to improve internal code quality?

It's to improve the internal code quality really, needs discussing on what is and isn't going to be made explicit, but going to wait until 3.2 is released as this is made for 4.0

**Comment by reduz** [comment]
I think this should be closed, as the move to C++17 (#36457) solved all the problems we were having thanks to guaranteed copy elision.

**Comment by neikeq** [comment]
@reduz This change is to ensure constructors that are meant to be called explicitly can never called implicitly by accident.

**Comment by Calinou** [comment]
It's generally considered good C++ practice to use `explicit` for constructors whenever possible: https://en.cppreference.com/w/cpp/language/explicit

**Comment by akien-mga** [comment]
This hasn't been updated for a while and didn't compile back then, so closing.

It might still be a change worth doing, this might need to be further assessed in a proposal and redone from scratch on the current `master` branch.

**Comment by mhilbrunner** [comment]
Yeah, this needs to be decided upon and reworked if we do it. Still, thanks for contributing and sorry we let this get stale due to the huge PR backlog!

**Pull request closed without merging**